File: value/GPTD.py
import numpy as np
import gym
import GPy
from copy import deepcopy


class GPTD:
    def __init__(self, nu, sigma0, gamma, kernel):
        
        self.nu = nu
        self.gamma = gamma
        self.sigma0 = sigma0
        self.kernel = kernel
        self.D = np.array([[]])
        self.A = np.array([[]])
        self.H_ = np.array([[]])
        self.Q_ = np.array([[]])
        self.K_ = np.array([[]])
        self.K_inv = np.array([[]])
        self.alpha_ = np.array([[]])
        self.C_ = np.array([[]])
        # self.r = np.array([[]])

    def k_(self,x):

        if (len(x.shape)==1):
            x = x[:,np.newaxis]
        assert (len(x.shape)==2 and x.shape[1]==1), "Check state dimensions"

        k_ = []
        for d in self.D:
            k.append(self.kernel(d,x)[0])

        return np.array(k_)

    def update(self, xt, xt_1, r):
        """
        xt and xt_1 are numpy arrays of dimension dx1
        r is a scalar reward
        """
        if (self.D.shape[0]==1 and self.D.shape[1]==0):
            # Empty dictionary
            self.D = np.zeros(xt.shape[0],2)
            self.D[:,0] = xt_1[:,0]
            self.D[:,1] = xt[:,0]
            K_t = np.zeros(2,2)
            K_t[:,0] = self.k_(xt_1)[:,0]
            K_t[:,1] = self.k_(xt)[:,0]
            K_t_inv = np.linalg.inv(K_t)
            At = np.eye(2)
            H_t = np.dot(np.array([[1,-self.gamma]]),self.A)
            Q_t = np.linalg.inv(np.dot(H_t,np.dot(K_t,H_t.T))+np.diag([self.sigma0**2]))
            alpha_t = np.dot(H_t,Q_t)*r
            C_t = np.dot(H_t.T,np.dot(Q_t,H_t))

        else:

            K_t_1 = deepcopy(self.K_)
            K_t_1_inv = deepcopy(self.K_inv)
            alpha_t_1 = deepcopy(self.alpha_)
            C_t_1 = deepcopy(self.C_)
            Q_t_1 = deepcopy(self.Q_)
            At_1 = deepcopy(self.A)
            H_t_1 = deepcopy(self.H_)

            # Compute et
            k_t_1 = self.k_(xt_1)
            k_t = self.k_(xt)
            ktt = self.kernel(xt,xt)
            at_1 = np.dot(K_t_1_inv,k_t_1)
            at = np.dot(K_t_1_inv,k_t)
            et = ktt - np.dot(k_t.T,at)
            et = et[0][0]
            
            delk_t_1 = k_t_1 - self.gamma*k_t
            assert (delk_t_1==np.dot(K_t_1,h_t)), "Check delk_t_1"

            gt = np.dot(Q_t_1,np.dot(H_t_1,delk_t_1))

            if (et > self.nu):
                # If et > nu
                # Update D
                D = np.zeros(self.D.shape[0],self.D.shape[1]+1)
                D[:,:-1] = self.D
                D[:,-1] = xt[:,0]
                self.D = D
                print("Dictionary updated")


                # Compute alphat and Ct
                c_t = np.dot(H_t_1,gt) - at_1
                delktt = np.dot(at_1.T,(delk_t_1 - self.gamma*k_t_1)) + self.gamma**2*ktt
                s_t = self.sigma0**2 + delktt - np.dot(delk_t_1.T,np.dot(C_t_1,delk_t_1))

                K_t = np.zeros(K_t_1.shape[0]+1,K_t_1.shape[1]+1)
                K_t[:-1,:-1] = K_t_1
                K_t[-1,-1] = ktt
                K_t[-1,:-1] = k_t.T
                K_t[:-1,-1] = k_t[:,0]

                K_t_inv = np.zeros(K_t_1_inv.shape[0]+1,K_t_1_inv.shape[1]+1)
                K_t_inv[:-1,:-1] = et*K_t_1_inv+np.dot(at,at.T)
                K_t_inv[-1,-1] = 1
                K_t_inv[-1,:-1] = -at.T
                K_t_inv[:-1,-1] = -at[:,0]
                K_t_inv = 1/et*K_t_inv

                alpha_t = np.zeros(alpha_t_1.shape[0]+1,alpha_t_1.shape[1])
                alpha_t[:-1,:] = alpha_t_1 + c_t/s_t*(np.dot(delk_t_1.T,alpha_t_1)-r)
                alpha_t[-1,:] = self.gamma/st*(np.dot(delk_t_1.T,alpha_t_1)-r)
                
                C_t = np.zeros(C_t_1.shape[0]+1,C_t_1.shape[1]+1)
                C_t[:-1,:-1] = C_t_1 + 1/s_t*np.dot(c_t,c_t.T)
                C_t[-1,-1] = self.gamma**2/s_t
                C_t[-1,:-1] = self.gamma/s_t*c_t.T
                C_t[:-1,-1] = self.gamma/s_t*c_t[:,0]
                
                Q_t = np.zeros((Q_t_1.shape[0]+1,Q_t_1.shape[1]+1))
                Q_t[:-1,:-1] = s_t*Q_t_1 + np.dot(gt,gt.T)
                Q_t[-1,-1] = 1
                Q_t[-1,:-1] = -gt.T
                Q_t[:-1,-1] = -gt[:,0]
                Q_t = Q_t/s_t
                
                At = np.zeros((At_1.shape[0]+1,At_1.shape[1]+1))
                At[:-1,:-1] = At_1
                At[-1,-1] = 1
                
                H_t = np.zeros((H_t_1.shape[0]+1,H_t_1.shape[1]+1))
                H_t[:-1,:-1] = H_t_1
                H_t[-1,:-1] = at_1.T
                H_t[-1,-1] = -self.gamma

            else:
                # else
                # D unchanged
                # Compute alphat and Ct

                h_t = at_1 - self.gamma*at
                ct = np.dot(H_t_1,gt) - h_t
                st = self.sigma0**2 - np.dot(ct.T,delk_t_1)
                
                K_t = K_t_1
                K_t_inv = K_t_1_inv

                alpha_t = alpha_t_1 + ct/st*(np.dot(delk_t_1.T,alpha_t_1) - r)
                
                C_t = C_t_1 + 1/st*np.dot(ct,ct.T)
                
                Q_t = np.zeros((Q_t_1.shape[0]+1,Q_t_1.shape[1]+1))
                Q_t[:-1,:-1] = st*Q_t_1 + np.dot(gt,gt.T)
                Q_t[-1,-1] = 1
                Q_t[-1,:-1] = -gt.T
                Q_t[:-1,-1] = -gt[:,0]
                Q_t = Q_t/st
                
                At = np.zeros((At_1.shape[0]+1,At_1.shape[1]))
                At[:-1,:] = At_1
                At[-1,:] = at.T
                
                H_t = np.zeros((H_t_1.shape[0]+1,H_t_1.shape[1]))
                H_t[:-1,:] = H_t_1
                H_t[-1,:] = h_t.T

        self.K_ = K_t
        self.K_inv = K_t_inv
        self.alpha_ = alpha_t
        self.C_ = C_t
        self.Q_ = Q_t
        self.A = At
        self.H_ = H_t
        

File: matlab/GPTD.m
classdef GPTD < handle

    properties (SetAccess='private')
        env
        nu
        gamma_
        sigma0
        sigmak
        D
        A
        H_
        Q_
        K_
        K_inv
        alpha_
        C_
    end
    
    methods
        function gptd = GPTD(env, nu, sigma0, sigmak, gamma_)
            gptd.env = env;
            gptd.nu = nu;
            gptd.sigma0 = sigma0;
            gptd.sigmak = sigmak;
            gptd.gamma_ = gamma_;
            gptd.D = [];
            gptd.A = [];
            gptd.H_ = [];
            gptd.Q_ = [];
            gptd.K_ = [];
            gptd.K_inv = [];
            gptd.alpha_ = [];
            gptd.C_ = [];
        end
        
        function k = kernel(gptd,x,y)
            % x,y are nx1 inputs that represent states
            k = exp(-sum((x-y).^2,1)/(2*gptd.sigmak^2));
        end
        
        function k_ = kernel_vector(gptd,x)
            if (size(x,1)==1)
                x = x';
            end
            if (size(x,1)~=size(gptd.D,1))
                disp('Check dimension');
                return;
            end
            k_ = (gptd.kernel(gptd.D,repmat(x,1,size(gptd.D,2)))');
        end
        
        function update(gptd, xt, xt_1, r, gamma_)
            if (size(xt,1)==1)
                xt=xt';
            end
            if (size(xt_1,1)==1)
                xt_1=xt_1';
            end
            if isempty(gptd.D)
                gptd.D(:,1) = xt_1;
                %gptd.D(:,2) = xt;
                K_t = gptd.kernel(xt_1,xt_1); ;
                K_t_inv = inv(K_t);
                at =  1;
                h_t = 1;
                Q_t =  ([1/(h_t*K_t*h_t' + gptd.sigma0^2)]);
                alpha_t = [0]; %H_t'*Q_t*r;
                C_t = 0; %H_t'*Q_t*H_t;
                
                gptd.K_ = K_t;
                gptd.K_inv = K_t_inv;
                gptd.alpha_ = alpha_t;
                gptd.C_ = C_t;
                gptd.A = at;
                gptd.H_ = h_t;
            end
            K_t_1 = gptd.K_;
            K_t_1_inv = gptd.K_inv;
            alpha_t_1 = gptd.alpha_;
            C_t_1 = gptd.C_;
            At_1 = gptd.A;
            h_t_1 = gptd.H_;

            k_t_1 = gptd.kernel_vector(xt_1);
            k_t = gptd.kernel_vector(xt);
            ktt = gptd.kernel(xt,xt);
            at = K_t_1_inv*k_t;
            et = ktt - k_t'*at;

            delk_t = k_t_1 - gamma_*k_t;

            if ((abs(et) - gptd.nu) > 10^(-5))
                gptd.D(:,size(gptd.D,2)+1) = xt;
                % Dimension issues
                %c_t = H_t_1'*gt - at_1;
                %
                %delktt = at_1'*(delk_t_1 - gamma_*k_t) + gamma_^2*ktt;
                %s_t = gptd.sigma0^2 + delktt - delk_t_1'*C_t_1*delk_t_1;

                K_t = [K_t_1,k_t;k_t',ktt]; 

                K_t_inv = [et*K_t_1_inv+at*at' -at ; -at' 1]/et;
                at = [zeros(size(At_1, 1),1); 1];
                h_t = [At_1; -gamma_];
                delk_tt = At_1'*(k_t_1 - 2*gamma_*k_t) + gamma_^2*ktt;
                ct = h_t - [C_t_1*delk_t; 0];
                dt = r - delk_t'*alpha_t_1;
                st = delk_tt - delk_t'*C_t_1*delk_t;
                alpha_t_1 = [alpha_t_1; 0];
                C_t_1 = [C_t_1 zeros(size(C_t_1,1),1) ; zeros(size(C_t_1,2),1)', 0];
                gptd.K_ = K_t;
                gptd.K_inv = K_t_inv;
                gptd.A = at;

            else
%                 if (~any(gptd.D==xt))
%                     disp(strcat('Why should it be excluded? ',int2str(xt),', error ',string(et)));
%                 end
                h_t = At_1 - gamma_*at;
                delk_tt = h_t'*delk_t;
                dt = r - delk_t'*alpha_t_1;
                ct = h_t - C_t_1*delk_t;
                st = delk_t'*ct;
            end
            alpha_t = alpha_t_1 + ct*dt/st;
            C_t = C_t_1 + ct*ct'/st;

            if (any(isnan(alpha_t)))
                disp('Check alpha!');
            end

            gptd.alpha_ = alpha_t;
            gptd.C_ = C_t;
            gptd.H_ = h_t;
        end

        function build_posterior_monte_carlo(gptd, policy, num_episodes, max_episode_length, debug_)
            s = gptd.env.reset();
            for e=1:1:num_episodes
                is_terminal = false;
                num_steps = 0;
                while ((~is_terminal) && (num_steps < max_episode_length))
                    num_steps = num_steps + 1;
                    a = policy(s);
                    [s_, r, is_terminal] = gptd.env.step(a);
                    gptd.update(s_, s, r, gptd.gamma_);
                    s = s_;
% Display grid world
%                     disp(strcat('Current : ',int2str(s)));
%                     V = gptd.get_value_function(1:1:gptd.env.num_states);
%                     V = reshape(V,4,4);
%                     imagesc(V);
%                     colorbar;
%                     close;

% Display for swing-up
                    if(mod(e,10) == 0 && num_steps==1)
                        dx = (gptd.env.x_limits(2)-gptd.env.x_limits(1))/(gptd.env.num_points_x - 1); 
                        dx_dot = (gptd.env.x_dot_limits(2)-gptd.env.x_dot_limits(1))/(gptd.env.num_points_x_dot - 1);
                        [grid_x,grid_x_dot] = meshgrid(gptd.env.x_limits(1):dx:gptd.env.x_limits(2),gptd.env.x_dot_limits(1):dx_dot:gptd.env.x_dot_limits(2));
                        V = gptd.get_value_function([reshape(grid_x, size(grid_x,1)*size(grid_x,2), 1), reshape(grid_x_dot, size(grid_x,1)*size(grid_x,2), 1)]');
    %                     if(max(V)>50)
    %                         keyboard
    %                     end
                        V = reshape(V, 51, 51);
                        x = [gptd.env.x_limits(1), gptd.env.x_limits(2)];
                        y = [gptd.env.x_dot_limits(1), gptd.env.x_dot_limits(2)];
                        figure, title(['Iteration ', num2str(e)])
                        imagesc(x,y,V);
                        xlabel('theta'); ylabel('theta-dot');
                        colorbar;
                        hold on;
                        scatter(gptd.D(1,:),gptd.D(2,:),'MarkerFaceColor',[1 0 0],'LineWidth',1.5);
                        pause(0.1)
                    end
                end
                s_ = gptd.env.reset();
                if (size(gptd.D,2)>gptd.env.num_states)
                    disp('Check dictionary size');
                end
                %gptd.update(s_, s, 0, 0);
                s = s_;
                if (debug_)
                    disp(strcat('Episode : ',int2str(e),', Dictionary size : ',int2str(size(gptd.D,2))));
                end
            end
        end
        
        function build_posterior_vi(gptd, states, policy, num_iterations, debug_)
            
            for iter = 1:1:num_iterations
                if (debug_)
                    disp(strcat('Iteration : ',int2str(iter)));
                end
                is_terminal = false;
                for s=1:1:size(states,2)
                    s1 = states(:,s);
                    if (is_terminal)
                        gptd.update(s1,s2,0,0);
                    end
                    if (gptd.env.set(s1))
                        continue;
                    end
                    [s2, r, is_terminal] = gptd.env.step(policy(s1));
                    gptd.update(s2,s1,r,gptd.gamma_);
                end 
            end
            
        end
        
        function V = get_value_function(gptd, states)
            V = zeros(size(states,2),1);
            for i=1:1:size(states,2)
                s = states(:,i);
                V(i) = gptd.kernel_vector(s)'*gptd.alpha_;
            end
        end
        
        function visualize(gptd, grid_x, grid_x_dot) % Assuming a 2D state-space... Make it general?
            states = zeros(2,size(grid_x,2)*size(grid_x_dot,1));
            for i = 1:1:size(grid_x,2)
                for j = 1:1:size(grid_x_dot,1)
                    states(:,(i-1)*size(grid_x_dot,1) + j) = [grid_x(j,i);grid_x_dot(j,i)];
                end
            end
            V = gptd.get_value_function(states);
            figure;
            x = [gptd.env.x_limits(1), gptd.env.x_limits(2)];
            y = [gptd.env.x_dot_limits(1), gptd.env.x_dot_limits(2)];
            imagesc(x,y,reshape(V,size(grid_x_dot,1),size(grid_x,2)));
            xlabel('theta'); ylabel('theta-dot');
            title('GPTD Value function');
            colorbar;
        end
    end
end